### ISSUE TITLE
Can't lookup element of result of dot-dot traversal

### ISSUE THREAD

**Issue by olivergierke** [jayway/JsonPath#248]
```java
String source = "{\"user\" : {\"firstname\" : \"Dave\", \"lastname\" : \"Matthews\" }}";

System.out.println(JsonPath.read(source, "$..firstname")); // prints ["Dave"]
System.out.println(JsonPath.read(source, "$..firstname[0]")); // prints []
```

The latter isn't expected. That basically prevents canonical lookups for fields expected to be contained anywhere in the document but exactly once.

**Comment by jlolling** [comment]
This is correct and not an error. firstnam is not an array, therefore the reference to an array cannot work.

I would not recognize this as an error.

Cheers
Jan

**Comment by olivergierke** [comment]
What's the fundamental difference to [this example](https://jsonpath.herokuapp.com/?path=$..book[2]) then? My fundamental goal is that I'd like to refer to the first field named `firstname` no matter where it appears in the document. That doesn't seem possible here.

In my case, I actually know there will only be one match, but I can't map the former expression to a `String` as it returns an array. So I thought I basically put the uniqueness knowledge into my JSONPath expression, but then end up with nothing found.

**Comment by olivergierke** [comment]
Oh, @nebhale's implementation behaves as expected: https://jsonpath.herokuapp.com/?path=$..category[0]#nebhale-tab

**Comment by olivergierke** [comment]
One more (hand probably final) thought: the `..` operator turns the expression into an array. If I evaluate `$.books` and `$..category` in the example they both evaluate to an instance of `JSONArray` — the former from the document structure, the latter because of the operator applied.  So I'd argue it's inconsistent that I can apply the subscript operator to the former but not the latter.
